### ISSUE TITLE
How to use this model for Image Retrieval?

### ISSUE THREAD

**Issue by Catstyle** [NVlabs/PS3#10]
Hi, thanks for this wonderful work.
I am currently working on a Image Retrieval service and found this newly released PS3 models/weights.
I have a few questions.

1. From the QuickStart examples, i noticed that the outputs contains multiple scale and each one with lots features.
How to get the final features from [B, C, H, W] for the whole image? (For siglip2, i get [1, 1152] embeddings for the whole image)

2. Does text_model's output suitable for T2I task?

3. Are these models Multilingual?

Thanks for your reply.

**Comment by BIGBALLON** [comment]
Some question  here

**Comment by bfshi** [comment]
Hi @Catstyle and @BIGBALLON,

1. To get the feature map of the image, you can use the built-in `format_features_into_feature_maps` function, as described [here](https://github.com/NVlabs/PS3#4-format-the-encoded-features-into-masked-feature-maps). Note that the feature map here is multi-scale, i.e., this function will return a list of feature maps where each feature map comes from each scale of the image.

2. We didn't test any image retrieval task in the paper, but theoretically it should work because the pre-trained SigLIP can do retrieval pretty well and we continue pre-training PS3 on top of that, and we also use the same SigLIP loss objective. Using PS3 for image retrieval should be pretty straightforward:

```
from PIL import Image
from ps3 import PS3VisionModel, PS3ImageProcessor, PS3Tokenizer, PS3TextModel

# Load the PS3 model and processor.
vision_model = PS3VisionModel.from_pretrained("nvidia/PS3-1.5K-SigLIP")
processor = PS3ImageProcessor.from_pretrained("nvidia/PS3-1.5K-SigLIP")
vision_model.cuda().eval()
tokenizer = PS3Tokenizer.from_pretrained("nvidia/PS3-1.5K-SigLIP")
text_model = PS3TextModel.from_pretrained("nvidia/PS3-1.5K-SigLIP")
text_model.cuda().eval()

# You can replace it with your own image.
image = Image.open("assets/test_images/dock.jpg")

# Preprocess the image.
x = processor(image)["pixel_values"][0].unsqueeze(0).cuda()

# Extract high-res vision feature
vision_outs = vision_model(x, num_look_close='all', output_hidden_states=False)
vision_feature = vision_outs.pooled_output
print(features.shape)  # (1, 1152)

# Extract text embedding
text = ["A tall spire with a cross at the top of the building."]
text = tokenizer(text).cuda()
text_feature = text_model(text).pooled_output  # (1, 1152)
```

Then you can use `vision_feature` and `text_feature` for retrieval.

3. The base SigLIP we used here is only trained on English (but one can also use mSigLIP, a multilingual version of SigLIP, as the base model in PS3 pre-training), and our data only has English, so it's not multilingual.

**Comment by Catstyle** [comment]
Thanks. I will take a look into it.
